### ISSUE TITLE
Improve parser error formatting

### ISSUE THREAD

**Issue by goto-bus-stop** [apollographql/apollo-rs#365]
ref https://github.com/apollographql/apollo-rs/pull/347#discussion_r1020325156

Right now parser errors show the index and the current token it's looking at, which isn't all that helpful to end users. IIRC Simon mentioned using the Python REPL to copy-paste the GraphQL source and slice it at the error index to get the context he needed.

Maybe parser errors can use miette, like other parts of the project. We could even give suggestions in the future (like if we throw an "incomplete spread operator" error, we can suggest replacing it by a complete spread operator).

**Comment by goto-bus-stop** [comment]
we also talked about letting the router report detailed errors in a machine-readable format (i.e. JSON)

the graphql spec describes an error format:
https://spec.graphql.org/October2021/#sec-Errors.Request-errors

we could add a conversion to json in that shape (or probably a struct that impls Serialize) for parse errors and validation errors.

**Comment by erikwrede** [comment]
Hey everyone, 
more summarized machine-readable details would definitely be helpful, especially those necessary to output spec compliant errors!

Is it already possible to extract the necessary information for a spec compliant format from the `db.ast(file_id)` query, or is this still a userland task? More specifically, the `location: line, column` part. As far as I could tell, the `path` field still has to be generated by traversal.

It might help to add the original parser syntax error to the `SyntaxError` `ApolloDiagnostic`, so the original info is available to consumers of functions like `validate` without having to traverse the AST again.

**Comment by goto-bus-stop** [comment]
The `line`, `column` is only available from the `ApolloDiagnostic::SyntaxError` at the moment (through miette, i'm actually not sure from memory if it has an API that exposes this). diagnostics IIRC don't have file paths at all at the moment. I want to address that very soon, this week or the next. It would likely use file IDs, and then you can use `db.input(file_id).filename()` to look up the name.

**Comment by erikwrede** [comment]
Yes, we're using the compiler since we plan to use it for validation. Currently working on converting the parsed documents back to a GQL-Python compatible format and representing the errors. Since the issue mentioned GQL-Spec-style errors I thought this would be a good place to comment 🙂

Using `thiserror` sounds great as it interfaces nicely with `miette`, preserving those fancy console outputs while making the errors easier to process. Looking forward to seeing what you come up with ☺️
